FF119 Add PublicKeyCredential.parseCreationOptionsFromJSON() (#29436)
mdn/content: files/en-us/web/api/publickeycredential/parsecreationoptionsfromjson_static/index.md

---
title: "PublicKeyCredential: parseCreationOptionsFromJSON() static method"
short-title: parseCreationOptionsFromJSON()
slug: Web/API/PublicKeyCredential/parseCreationOptionsFromJSON_static
page-type: web-api-static-method
status:
  - experimental
browser-compat: api.PublicKeyCredential.parseCreationOptionsFromJSON_static
---

{{APIRef("Web Authentication API")}} {{SeeCompatTable}}{{securecontext_header}}

> The **`parseCreationOptionsFromJSON()`** static method of the {{domxref("PublicKeyCredential")}} interface converts a particular JSON-friendly representation of the [`publicKey` create credentials options object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure) into the `publicKey` structure.

Author: FYI I have used the term "JSON-friendly representation". The spec uses the term "JSON type representation". I prefer friendly but can change to the spec term if you think is better.

Reviewer: JSON string?

The method is a convenience function for deserializing information provided by a relying server to a web app in order to create a credential.

## Syntax

```js-nolint
PublicKeyCredential.parseCreationOptionsFromJSON(options)
```

### Parameters

- `options`

  - : An object with the same structure as the Web Authentication API [`publicKey` create credentials options object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure), but with [base64url](/en-US/docs/Glossary/Base64) encoded strings used in place of buffer properties.

### Return value

> An object with the Web Authentication API [`publicKey` create credentials options object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure).

Author: FYI this term gets used a lot in this doc. It's a bit of a mouthful but it is a precise way to refer to this particular object structure, compared to others. The spec uses a dictionary name.

> `publicKey` create credentials options object structure`

The web app passes this information to an authenticator to create the credential, by calling [`navigator.credentials.create()`](/en-US/docs/Web/API/CredentialsContainer/create) with an argument that contains the server-supplied data in the [`publicKey` create credentials options object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure).

The specification does not define how the information needed for creating a credential is sent.
A convenient approach is to first encapsulate the information in an object that mirrors the structure of the [`publicKey` create credentials options object](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure), encoding buffer properties such as the `challenge` and `user.id` as [base64url](/en-US/docs/Glossary/Base64) strings.
This object can be serialized to a [JSON](/en-US/docs/Glossary/JSON) string and then deserialized by the web app to the [`publicKey` create credentials options object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure) using **`parseCreationOptionsFromJSON()`**.

## Examples

> A server might create JSON credentials using something like this (taken from the [`AuthenticatorResponse` example "getting an AuthenticatorAttestationResponse"](/en-US/docs/Web/API/AuthenticatorResponse#getting_an_authenticatorattestationresponse)):

Author: ```suggestion
A server might create the options object required for creating new credentials using code that looks something like this (taken from the [`AuthenticatorResponse` example "getting an AuthenticatorAttestationResponse"](/en-US/docs/Web/API/AuthenticatorResponse#getting_an_authenticatorattestationresponse)):
```

```js

> const publicKey = {

Author: ```suggestion
const credentialCreateOptions = {
```

  challenge:
    "21, 31, 105, " /* 29 more random bytes generated by the server in this string */,
  rp: {
    name: "Example CORP",
    id: "login.example.com",
  },
  user: {
    id: "16",
    name: "canand@example.com",
    displayName: "Carina Anand",
  },
  pubKeyCredParams: [
    {
      type: "public-key",
      alg: -7,
    },
  ],
};

> //Stringify to serialize/send
> const publicKeyJSON = JSON.stringify(publicKey);

Author: ```suggestion
// Stringify to serialize/send
const credentialCreateOptionsJSON = JSON.stringify(credentialCreateOptions);
```

```

> The `publicKeyJSON` can be sent to the web app, which will parse the JSON back into an object.

Author: ```suggestion
The `credentialCreateOptionsJSON` string can be sent to the web app, which will parse the JSON back into an object.
```

> Below we show how you might use `parseCreationOptionsFromJSON()` to convert this parsed object "`receivedPublicKeyJSON`" to the correct form to use in `navigator.credentials.create()`:

Author: ```suggestion
Below we will name this "reconstituted" object `receivedCreateCredentialOptions` (though it is the same as `createCredentialOptions` on the server). 
We use `parseCreationOptionsFromJSON()` to convert this to the correct form to use in `navigator.credentials.create()`:
```

```js
const publicKey = PublicKeyCredential.parseCreationOptionsFromJSON(

>   receivedPublicKeyJSON,

Author: ```suggestion
  receivedCreateCredentialOptions,
```

This object can be serialized to a [JSON](/en-US/docs/Glossary/JSON) string and then deserialized by the web app to the [`publicKey` create credentials options object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure) using **`parseCreationOptionsFromJSON()`**.

## Examples

A server might create the options object containing information required (by the web app/authenticator) for creating new credentials, using code that looks something like this (taken from the [`AuthenticatorResponse` example "getting an AuthenticatorAttestationResponse"](/en-US/docs/Web/API/AuthenticatorResponse#getting_an_authenticatorattestationresponse)):

```js
const createCredentialOptions = {
  challenge:
    "21, 31, 105, " /* 29 more random bytes generated by the server in this string */,
  rp: {
    name: "Example CORP",
    id: "login.example.com",
  },
  user: {
    id: "16",
    name: "canand@example.com",
    displayName: "Carina Anand",
  },
  pubKeyCredParams: [
    {
      type: "public-key",
      alg: -7,
    },
  ],
};

// Stringify to serialize/send
const createCredentialOptionsString = JSON.stringify(createCredentialOptions);
```

> The `createCredentialOptionsString` string can be sent to the web app and parsed back from a JSON-formatted string into a JavaScript object.

Author: From https://github.com/mdn/content/pull/29436/files#r1350111225

> Ah, so you are saying that you'd create the credentials object on a separate server, then stringify it and send it to the client, then use parseCreationOptionsFromJSON() to turn it into the appropriate format object to then pass into a create() call.
> 
> This has got me thinking about why you might do this, rather than just creating the create() parameter on the client...I suppose it is feasible that you might have some separate server that handles creating the credentials.
> 
> I am just wondering if you should explain the flow here a bit more explicitly, for others that might have similar questions.

You'd think English was a second language for me. I have rewritten this! (which is why your note is reposted). The essential change is that the server creates options, not credentials.

The options for creating a credential are created on the server. These are stringified, sent to the web app, parsed back to the same object - the web app probably does this somehow as a `fetch()` call and gets the object back as `Response.json()`.

The format of the options object on the server, over the wire and unparsed on the web app is not defined in the spec, but if you define it to be "like the format expected by `create()` except using strings instead of buffers then what arrives at the web app can be converted using this parsing function and passed to `create()`.

To be very clear, the authenticator always creates the credentials. Not the server. Though in theory there is no reason "some server" could not have this role.

Reviewer: Cool, this makes more sense now.

> Below we will name this "reconstituted" object `receivedCreateCredentialOptions` (though it is the same as `createCredentialOptions` on the server). 

Author: ```suggestion
Below we will name this "reconstituted" object `receivedCreateCredentialOptions` (though it is the same as `createCredentialOptions` on the server). 
```
```suggestion
Below we will name this "reconstituted" object `createCredentialOptionsJSON` (though it is the same as `createCredentialOptions` on the server). 
```

We use `parseCreationOptionsFromJSON()` to convert this to the correct form to use in `navigator.credentials.create()`:

```js

> const publicKey = PublicKeyCredential.parseCreationOptionsFromJSON(

Author: ```suggestion
const createCredentialOptions = PublicKeyCredential.parseCreationOptionsFromJSON(
```

  receivedCreateCredentialOptions,
);

navigator.credentials

>   .create({ publicKey })

Author: ```suggestion
  .create({ createCredentialOptionsJSON })
```

A convenient approach is for the server to encapsulate the information in a {{glossary("JSON type representation")}} of the [`publicKey` create credentials options object](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure) that mirrors its structure but encodes buffer properties such as the `challenge` and `user.id` as [base64url](/en-US/docs/Glossary/Base64) strings.
This object can be serialized to a [JSON](/en-US/docs/Glossary/JSON) string, sent to the web app and deserialized, and then converted to the [`publicKey` create credentials options object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure) using **`parseCreationOptionsFromJSON()`**.

## Examples

When registering a new user, a relying party server will supply information about the expected credentials to the web app.
The code below defines this information in the form described in the [`options` parameter](#options) above (taken from the ["getting an AuthenticatorAttestationResponse"](/en-US/docs/Web/API/AuthenticatorResponse#getting_an_authenticatorattestationresponse) in `AuthenticatorResponse`):

```js

> const createCredentialOptionsJSON = {

Author: Note, new naming convention to mirror spec and provide consistency. `createCredentialOptionsJSON` means "JSON representation. Without the JSON it is the original object. The example now shows this same object on both the server and the web app, which is much more clear.

  challenge:
    "21, 31, 105, " /* 29 more random bytes generated by the server in this string */,
  rp: {
    name: "Example CORP",
    id: "login.example.com",
  },
  user: {
    id: "16",
    name: "canand@example.com",
    displayName: "Carina Anand",
  },
  pubKeyCredParams: [
    {
      type: "public-key",
      alg: -7,
    },
  ],
};
```

Because this object only uses JSON data types, it can be be serialized to JSON using [`JSON.stringify()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) and sent to the web app.

```js
JSON.stringify(createCredentialOptionsJSON);

> ```

Author: So in real live I _think_ that probably the web app requests to authenticate, so it would do a get request for the options, and the server would respond with a response object that contains this JSON. I have no real example, so this is the simple version that is harder to screw up.

> The web app can deserialize the JSON string back to a `createCredentialOptionsJSON` object (not shown).

Author: "not shown" because this depends on how the data is sent. Assuming it is requested, the Response object would have a .json object that would be the deserialized object.

---
title: "PublicKeyCredential: parseCreationOptionsFromJSON() static method"
short-title: parseCreationOptionsFromJSON()
slug: Web/API/PublicKeyCredential/parseCreationOptionsFromJSON_static
page-type: web-api-static-method
browser-compat: api.PublicKeyCredential.parseCreationOptionsFromJSON_static
---

> 

Author: @chrisdavidmills This is the first static method I have to create here. Would you be able to have a look - it will be the pattern for the other two. My main concern is whether you think this is sufficient 
- do I need to expand out the parameters and return value. It makes little sense to me, since the values are the same except buffers are base64 encoded.
- Is the example sufficient? It is very basic. I could instead of showing what the server sends, do a live example and show the actual converted value. But I think this shows the main point.

Reviewer: LGTM; the structure seems fine, and the level of detail is OK.

{{APIRef("Web Authentication API")}}{{securecontext_header}}

> The **`parseCreationOptionsFromJSON()`** static method of the {{domxref("PublicKeyCredential")}} interface deserializes a correpsonding JSON object into the [`publicKey` object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure).
> 
> Typically a server will create public key credential options, including a challenge that the client must sign and return in order to ensure the integrity of the registration process, serialize them to JSON, and then send them to the client.
> The client can then use this method to deserialize the JSON before passing it to [`navigator.credentials.create()`](/en-US/docs/Web/API/CredentialsContainer/create) in order to create the [Web Authentication API](/en-US/docs/Web/API/Web_Authentication_API) credentials.
> 
> The JSON object to be deserialized shares the same structure as the [`publicKey` object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure) passed to `navigator.credentials.create()`, except that [base64url](/en-US/docs/Glossary/Base64) strings are used in attributes in place of buffers (for example in the `challenge`, and in the `user.id`).

Author: First thoughts on fixing this. @chrisdavidmills I'm still working on this, but does it capture the essence?

```suggestion
The **`parseCreationOptionsFromJSON()`** static method of the {{domxref("PublicKeyCredential")}} interface converts a JSON-friendly representation of the [`publicKey` object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure) into that structure.

The object to be converted must have the same structure as the [`publicKey` object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure) passed to `navigator.credentials.create()`, except that [base64url](/en-US/docs/Glossary/Base64) strings are used in attributes in place of buffers (for example in the `challenge`, and in the `user.id`).

The method may be used in the first step of the [registration process](LINK TO WEB AUTHN API docs section) to convert information about credential requirements from the server into the form in which they can be passed to the `create()` API. 
The server is not required to send the information in any particular format, but using the JSON friendly representation simplifies serialization using `JSON.stringify()`, deserialization using `JSON.parse()` and then conversion to the format expected by `create()`.
```

Author: Decided on alternative wording.

## Syntax

```js-nolint
PublicKeyCredential.parseCreationOptionsFromJSON(options)
```

### Parameters

- `options`

>   - : A JSON object with the same structure as the Web Authentication API the [`publicKey` object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure).

Reviewer: ```suggestion
  - : A JSON object with the same structure as the Web Authentication API [`publicKey` object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure).
```

Reviewer: Also, You should be careful with the term "JSON object"; purists argue that there isn't really any such thing and, looking at the example, you are passing a string into the method?

I couldn't find the spec for this method to check what data type it expects, but I'd encourage you to check it.

Author: Excellent point thank you (and for your review).

I'm having a lot of trouble with the [Web Authentication API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Authentication_API) doc because it uses terms without defining them. What's a relying server? Whats an MFA? I _think_ I know, but perhaps it is some kind of third party server, or perhaps it is the website that hosts the web app. Anyway,

I _think_ it is like this:

- Web app - the web page code that wants to authenticate with a backend server
- Relying server - the server that is supporting the web app code. The "relying" means "relying on information from an authenticator"?
- Authenticator - something accessible (or in) the user agent that can create and store credentials. This might be code, or a physical stick. This is what the web app interacts with.

So basically if someone wants to log into the web app they need to create credentials. They start by doing:

> 1. The relying party server sends user and relying party information to the web app handling the registration process, along with the "challenge", using an appropriate secure mechanism (for example [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) or [Fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)).

So I guess there is a step 0 where the user kicks this off, perhaps by going to a specific URL?

Anyway,
- the web app gets  info from the server that allows it to create a credentials object, that gets passed to the authenticator when you do the registration process. 
- The authenticator generates keys and stores them, signs and returns the challenge along with the public key and possibly other info. 
- The web app sends that to the server.
- The relying server can store the public key. It can also check that the challenge that has been returned was signed with the corresponding private key. At this point you've securely shared keys and identity.

Later on if the web app visits the site the server can send a challenge specifying the kind of response it wants back (location where it expects the authentication info to be). The web app can pass this to the authenticator using a get call, which then returns a credentials response that has been signed using stored key. The server can validate this and the challenge using its stored public key. 

About right?

With that context, I can tell you that the JSON objects referred to here are JavaScript Objects that act as  JSON representations of the credentials objects that need to exchanged with the server. So basically they are objects that you can call JSON.stringify() and parse() on to send over the wire. You then end up with these objects which are almost the same as are used by the Web Authn API - and the methods in this PR do the final conversion to change encoded strings to buffers.

How I word that, I do not know!

Reviewer: So, let's get started here.

- relying server: I don't really use the term "relying server". I more generally use the term "relying party" to refer to the company (and their app) that is using WebAuth to provide sign-in functionality for their app/service. I do link the first instance to the Wikipedia page about it, but maybe we should create a brief glossary page to define it, and link to that instead. It is also very relevant for other related APIs such as FedCM, in which the relying parties rely on federated auth services for sign-in to their own service.
- MFA: Multi-factor authentication. Authentication that relies on not a single factor (usually just username and password), but multiple factors (e.g. password and authenticator, or password and OTP via an SMS) to sign-in. I do expand the first instance of this on the WebAuthn page, but again, maybe we need a glossary entry to clearly define it. I think I kind of just assumed it was obvious, which in hindsight is a dangerous game to play.

Reviewer: On to your definitions:

> Web app - the web page code that wants to authenticate with a backend server

Yeah, kind of. When I say app in the WebAuthn doc, I am referring to the app that wants to provide secure sign-in via the WebAutn API and its associated back-end machinery.

> Relying server - the server that is supporting the web app code. The "relying" means "relying on information from an authenticator"?

Not really. See previous comment.

> Authenticator - something accessible (or in) the user agent that can create and store credentials. This might be code, or a physical stick. This is what the web app interacts with.

Sort of. This is the server or physical device that the WebAuthn API interacts with to generate the public key credential used to create a secure sign-in credential during a `create()` call, and to verify a previously created secure sign-in credential during a `get()` call.

> Attestation?

This is a really bloody difficult word to define. In all the research I did, I never found a concrete, clear, understandable definition of it, as it relates to authentication, hence the definitions of it on MDN not being very good. From what I have read, I get that it kinda means the mechanism by which the authenticator can prove that it is trustworthy and belongs to the platform it says it does, for an extra level of security during authentication.

Reviewer: > I think part of the problem for me here is that the examples and names call this credential information from the server a publicKey. Why? It's lots of things, but it isn't that.

The public key is the thing the authenticator creates during the `create()` flow and passes back to the web app in response to a successful `create()` call (the promise fulfills with a `PublicKeyCredential` object). The web app is then supposed to store this on the server along with the user credentials. When the user then tries to sign in, the `get()` flow then uses this stuff to verify the user is who they say they are.

Reviewer: > With that context, I can tell you that the JSON objects referred to here are JavaScript Objects that act as JSON representations of the credentials objects that need to be exchanged with the server. So basically they are objects that you can call JSON.stringify() and parse() on to send over the wire. You then end up with these objects which are almost the same as are used by the Web Authn API - and the methods in this PR do the final conversion to change encoded strings to buffers.

Makes sense. I would say something like "are JavaScript Objects representing the credentials that need to be exchanged with the server" and the methods "facilitate conversion of the objects to and from a JSON string as they are passed over the wire to/from client to server". Maybe?

Author: > * MFA: Multi-factor authentication. ... I think I kind of just assumed it was obvious, which in hindsight is a dangerous game to play.

I know the acronym, and I'd tend to agree with you that given the context it perhaps should be guessable. IN fact I just looked at your doc and you define the term in the very first paragraph. So I'm wrong on this one.

> So, let's get started here.
> 
> * relying server: I don't really use the term "relying server". I more generally use the term "relying party" to refer to the company (and their app) that is using WebAuth to provide sign-in functionality for their app/service. I do link the first ...

OK, so we basically have a server that we refer to as the relying party. This "provides access to a secured software", which is our web app running in the browsers.

We don't know why it is called a "relying party". That's a bit irritating because if I knew why it was called that it might help create a framework for my understanding of what we're trying to achieve. But anyhow....

The web app talks to an authenticator, which is back end machinery. It is something that can create public/private keys and sign the challenge and such and associate them with a particular user. It is  accessible to the Web APP via the Web Authn API. I guess the authenticator could be anything as long as it can store those keys and respond to the Web Auth API as expected.

The authenticator stores the keys and gives the public key back to the web app, which in turn passes it to the relying party. The relying party (server) now has a public key that it can use to verify "messages" from a particular user and authenticate them.

That's really the heart of it - once the server has a public key associated with a particular user and their authentication system they can always authenticate that user with a challenge that is returned signed by that key.

I don't know why this has been so hard to get my head around. I'm not one to accept responsibility, so it must be the docs :-). 

I still feel that we're not helping ourselves with the naming  - it might help the experts to refer to the server as the "relying party", but it is no help to a beginner. The examples using public key as the name of the credential options is not helpful either. I think a picture showing server, app, authenticator might help. I'll leave with you though.

>I would say something like "are JavaScript Objects representing the credentials that need to be exchanged with the server" and the methods "facilitate conversion of the objects to and from a JSON string as they are passed over the wire to/from client to server". Maybe?

Hmmm. I think you might have something here. I'll think a bit more about this.

Author: It's really hard because that misses that the structure you have to use for these is predefined - they are literally JSON friendly versions of the original. [SOB].

Reviewer: > OK, so we basically have a server that we refer to as the relying party. This "provides access to a secured software", which is our web app running in the browsers.

> We don't know why it is called a "relying party". That's a bit irritating because if I knew why it was called that it might help create a framework for my understanding of what we're trying to achieve. But anyhow....

I think this is the one bit I've got left to comment on here. The relying party, afaiui, _is_ the web app that is using webauthn, or fedcm, to provide a secure sign in.

It is called the relying party because it relies on a service outside the app itself to handle the sign-in process. This bit feels fairly clear to me. I guess you could say "3rd party service rather than "service outside the app itself ", but I'd hesitate to say that because "3rd party" makes you think of a separate library or framework, and its not generally that.

Author: >It is called the relying party because it relies on a service outside the app itself to handle the sign-in process. 

I think that's reasonable. 

But if so, then a lot of the docs are even more confused - e.g. "The relying party generates a "challenge" and sends it to the user agent ". Actually I believe it is the service "separate to the app" that generates the challenge.  We can't mix the roles like this an expect the docs to be intelligible.

The solution here is to separate the app and the service, define their roles, define clear names that are used everywhere to refer to them. We can still state that these are running on the same server.

My understanding is that the service role is to create and maintain a relationship between a public key and a user and their access to other services that might be exposed by the web app.

The role of the web app is just to exchange messages between the service and the authenticator via the Web API.

The role of the authenticator is to verify the user when needed, to store a private/public key for that user associated with the service, and to return challenges signed with the key when needed.

Once the service has verified identity I guess some other part of the server then use the association with a user to grant rights.

Author: I'm not going there unless forced :-)

>     The only difference is that the JSON object use strings in place of buffers: for example in the `challenge`.

Reviewer: ```suggestion
    The only difference is that the JSON object uses strings in place of buffers: for example in the `challenge`.
```

### Return value

An object with the Web Authentication API [`publicKey` object structure](/en-US/docs/Web/API/CredentialsContainer/create#publickey_object_structure).
The attribute values will match the original JSON object passed into the method, except for buffers, which will be decoded from the corresponding [base64url](/en-US/docs/Glossary/Base64) encoded strings.

### Exceptions

- `EncodingError` {{domxref("DOMException")}}

>   - : Raised if any part of the JSON options cannot be converted into the corresponding JSON object passed to [`navigator.credentials.create()`](/en-US/docs/Web/API/CredentialsContainer/create).

Reviewer: I normally say "Thrown" instead of "Raised; seems to fit better with the JS syntax (`Throw`, etc.)

Author: This become complicated when you're working with promises, since thrown objects are rejected promises rather than caught. That said, we're not working with promises, and I prefer Thrown when I can be sure I'll get away with it :-)
```suggestion
  - : Thrown if any part of the JSON options cannot be converted into the corresponding JSON object passed to [`navigator.credentials.create()`](/en-US/docs/Web/API/CredentialsContainer/create).
```

## Examples

> Consider a server might create JSON credentials "something" like this (taken from the [`AuthenticatorResponse` example "getting an AuthenticatorAttestationResponse"](/en-US/docs/Web/API/AuthenticatorResponse#getting_an_authenticatorattestationresponse)):

Reviewer: ```suggestion
Consider a server might create JSON credentials something like this (taken from the [`AuthenticatorResponse` example "getting an AuthenticatorAttestationResponse"](/en-US/docs/Web/API/AuthenticatorResponse#getting_an_authenticatorattestationresponse)):
```
